File: scm-java/src/main/scheme/primitives/PrimitiveModuleExports.java
package scheme.primitives;

import java.util.ArrayList;
import java.util.Collections;

import scheme.Modules;
import scheme.Pair;
import scheme.Primitive;
import scheme.SourcePos;
import scheme.SchemeError;

public class PrimitiveModuleExports extends Primitive {
    private Modules modules;

    public PrimitiveModuleExports(Modules modules) {
        this.modules = modules;
    }

    @Override
    public String name() {
        return "%module-exports";
    }

    @Override
    public String info() {
        return "Syntax: (%module-exports module-name)\n" +
               "Library: (scm core)\n" +
               "Description: Returns an alphabetically sorted list of all symbols exported by the named module.\n" +
               "Example:\n" +
               "  (%module-exports '(scheme base))";
    }

    @Override
    public Object apply(SourcePos pos, Object[] arguments) {
        checkArgs(pos, arguments, 1, 1);

        var module = modules.getModule(Modules.asModuleName(arguments[0]));
        if (module == null) {
            throw new SchemeError(pos, name() + ": module ~a not found", arguments[0]);
        }
        var symbols = new ArrayList<String>(module.exports.keySet());
        Collections.sort(symbols);
        return Pair.list(symbols.toArray());
    }
}

File: scm-csharp/scheme/primitives/PrimitiveModuleExports.cs
namespace scheme;

public class PrimitiveModuleExports : Primitive
{
    private Modules modules;

    public PrimitiveModuleExports(Modules modules)
    {
        this.modules = modules;
    }

    public override string Name() => "%module-exports";

    public override string Info() =>
        "Syntax: (%module-exports module-name)\n" +
        "Library: (scm core)\n" +
        "Description: Returns an alphabetically sorted list of all symbols exported by the named module.\n" +
        "Example:\n" +
        "  (%module-exports '(scheme base))";

    public override object Apply(SourcePos? pos, object[] arguments)
    {
        CheckArgs(pos, arguments, 1, 1);
        var module = modules.GetModule(Modules.AsModuleName(arguments[0]));
        if (module == null)
        {
            throw new SchemeError(pos, Name() + ": module ~a not found", arguments[0]);
        }
        var symbols = module!.Exports.Keys.OrderBy(k => k).ToList();
        return Pair.List(symbols.Select(s => (object)Value.Intern(s)).ToArray());
    }
}
